### PR TITLE
pickle the AutoML object

### PR DESCRIPTION
**Pull request by sonichi**
@ekzhu I set self.file=None after close() is called for making the AutoML object picklable. Could you review it?

### CODE REVIEW

**Review by ekzhu — COMMENTED** [review]
Looks good to me. Only note is that there should be a version number for pickled object because the pickled object will be outdated as the class evolves overtime. So when you unpickle an serialized object you need to compare its version with the current running version of AutoML. See example: https://stackoverflow.com/questions/2361140/pickled-object-versioning

**Comment by sonichi** [comment]
> Looks good to me. Only note is that there should be a version number for pickled object because the pickled object will be outdated as the class evolves overtime. So when you unpickle an serialized object you need to compare its version with the current running version of AutoML. See example: https://stackoverflow.com/questions/2361140/pickled-object-versioning

@ekzhu thanks. Could you check the updated version?

File: flaml/automl.py [hunk 1]
```diff
@@ -1208,9 +1226,11 @@ def _select_estimator(self, estimator_list):
                 gap = search_state.best_loss - self._state.best_loss
                 if gap > 0 and not self._ensemble:
                     delta_loss = (search_state.best_loss_old - 
-                                 search_state.best_loss)
+                                 search_state.best_loss) or \
+                                     search_state.best_loss
                     delta_time = (search_state.total_time_used - 
-                                 search_state.time_best_found_old)
+                                 search_state.time_best_found_old) or (
+                                     search_state.total_time_used + 1e-10)
```
Comment (qingyun-wu) [inline flaml/automl.py:1233]: is this correct? in this case the 'real' delta_time, i.e., (search_state.total_time_used - search_state.time_best_found_old) is 0, do we really want to set delta_time to be search_state.total_time_used + 1e-10? shouldn't set it as 1e-10?

File: flaml/searcher/blendsearch.py [hunk 2]
```diff
@@ -143,20 +144,31 @@ def _init_search(self):
         self._deadline = np.inf
 
     def save(self, checkpoint_path: str):
-        save_object = (self._metric_target, self._search_thread_pool,
-            self._thread_count, self._init_used, self._trial_proposed_by,
-            self._ls_bound_min, self._ls_bound_max, self._result,
-            self._deadline)
+        save_object = self
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:147]: Does it mean to save the whole BlendSearch instance?

File: flaml/searcher/blendsearch.py [hunk 3]
```diff
@@ -526,3 +538,87 @@ def _create_condition(self, result: Dict) -> bool:
         return len(self._search_thread_pool) < 2
 
 
+def create_next(client):
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:541]: What is the type of the input client?

File: flaml/searcher/blendsearch.py (hunk 2 above)
Comment (sonichi) [reply flaml/searcher/blendsearch.py:147]: yes, to ensure all the states are saved.

File: flaml/searcher/blendsearch.py (hunk 3 above)
Comment (sonichi) [reply flaml/searcher/blendsearch.py:541]: not defined in this repo

File: flaml/automl.py (hunk 1 above)
Comment (sonichi) [reply flaml/automl.py]: Thanks. Fixed.

File: flaml/automl.py [hunk 4]
```diff
@@ -1208,9 +1226,11 @@ def _select_estimator(self, estimator_list):
                 gap = search_state.best_loss - self._state.best_loss
                 if gap > 0 and not self._ensemble:
                     delta_loss = (search_state.best_loss_old - 
-                                 search_state.best_loss)
+                                 search_state.best_loss) or \
+                                     search_state.best_loss
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/automl.py:1230]: As defined in the paper.

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**